code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,513 @@
+	returnable_list += span_bold("Be sure to read <a href=\"https://tgstation13.org/wiki/Revenant\">the wiki page</a> to learn more.")
+	return returnable_list
+
+/mob/living/basic/revenant/proc/set_random_revenant_name()
+	var/list/built_name_strings = list()
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "spirit_type"))
+	built_name_strings += " of "
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "adverb"))
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "theme"))
+	name = built_name_strings.Join("")
+
+/mob/living/basic/revenant/proc/on_baned(obj/item/weapon, mob/living/user)
+	SIGNAL_HANDLER
+	visible_message(
+		span_warning("[src] violently flinches!"),
+		span_revendanger("As [weapon] passes through you, you feel your essence draining away!"),
+	)
+	inhibited = TRUE
+	update_mob_action_buttons()
+	addtimer(CALLBACK(src, PROC_REF(reset_inhibit)), 3 SECONDS)
+
+/mob/living/basic/revenant/proc/reset_inhibit()
+	inhibited = FALSE
+	update_mob_action_buttons()
+
+/// Incorporeal move check: blocked by holy-watered tiles and salt piles.
+/mob/living/basic/revenant/proc/incorporeal_move_check(atom/destination)
+	var/turf/open/floor/step_turf = get_turf(destination)
+	if(isnull(step_turf))
+		return TRUE // what? whatever let it happen
+
+	if(step_turf.turf_flags & NOJAUNT)
+		to_chat(src, span_warning("Some strange aura is blocking the way."))
+		return FALSE
+
+	if(locate(/obj/effect/decal/cleanable/food/salt) in step_turf)
+		balloon_alert(src, "blocked by salt!")
+		reveal(2 SECONDS)
+		temporary_freeze(2 SECONDS)
+		return FALSE
+
+	if(locate(/obj/effect/blessing) in step_turf)
+		to_chat(src, span_warning("Holy energies block your path!"))
+		return FALSE
+
+	return TRUE
+
+//reveal, stun, icon updates, cast checks, and essence changing
+/mob/living/basic/revenant/proc/reveal(time)
+	if(QDELETED(src))
+		return
+	if(time <= 0)
+		return
+	revealed = TRUE
+	invisibility = 0
+	incorporeal_move = FALSE
+	if(!unreveal_time)
+		to_chat(src, span_revendanger("You have been revealed!"))
+		unreveal_time = world.time + time
+	else
+		to_chat(src, span_revenwarning("You have been revealed!"))
+		unreveal_time = unreveal_time + time
+	update_appearance(UPDATE_ICON)
+	orbiting?.end_orbit(src)
+
+/mob/living/basic/revenant/proc/temporary_freeze(time)
+	if(QDELETED(src))
+		return
+
+	if(time <= 0)
+		return
+
+	ADD_TRAIT(src, TRAIT_NO_TRANSFORM, REVENANT_STUNNED_TRAIT)
+
+	if(unstun_time <= 0)
+		balloon_alert(src, "can't move!")
+		unstun_time = world.time + time
+	else
+		balloon_alert(src, "can't move!")
+		unstun_time = unstun_time + time
+
+	update_appearance(UPDATE_ICON)
+	orbiting?.end_orbit(src)
+
+/mob/living/basic/revenant/proc/cast_check(essence_cost)
+	if(QDELETED(src))
+		return
+
+	var/turf/current = get_turf(src)
+
+	if(isclosedturf(current))
+		to_chat(src, span_revenwarning("You cannot use abilities from inside of a wall."))
```

Comment:
Maybe balloon alerts?

Reply:
i'm not sure how to condense that but i can look at that